File: dertetgpt-ios/DertetHarness/ViewModels/SettingsViewModel.swift
import Foundation
import Combine

@MainActor
final class SettingsViewModel: ObservableObject {
    private let store = SettingsStore.shared

    @Published var selectedProvider: ProviderType
    @Published var apiKey: String = ""
    @Published var model: String = ""
    @Published var baseUrl: String = ""
    @Published var systemPrompt: String = ""
    @Published var justSaved: Bool = false
    @Published var availableModels: [String] = []
    @Published var modelsLoading: Bool = false

    private var modelsTask: Task<Void, Never>?

    init() {
        let current = store.currentSettings
        self.selectedProvider = current.provider
        self.apiKey = current.apiKey
        self.model = current.model
        self.baseUrl = current.baseUrl
        self.systemPrompt = current.systemPrompt
        refreshModels()
    }

    func onProviderSelected(_ provider: ProviderType) {
        guard provider != selectedProvider else { return }
        selectedProvider = provider
        justSaved = false
        availableModels = []
        let s = store.settings(for: provider)
        apiKey = s.apiKey
        model = s.model
        baseUrl = s.baseUrl
        refreshModels()
    }

    func refreshModels() {
        modelsTask?.cancel()
        let provider = selectedProvider
        let key = apiKey
        let url = baseUrl
        modelsLoading = true
        modelsTask = Task { [weak self] in
            let models = await ModelsAPI.fetchAvailableModels(style: provider.apiStyle, baseUrl: url, apiKey: key)
            guard let self, !Task.isCancelled else { return }
            self.availableModels = models
            self.modelsLoading = false
        }
    }

    func save() {
        let finalModel = model.trimmingCharacters(in: .whitespaces).isEmpty ? selectedProvider.defaultModel : model
        let finalBaseUrl = baseUrl.trimmingCharacters(in: .whitespaces).isEmpty ? selectedProvider.defaultBaseUrl : baseUrl
        store.saveProviderConfig(
            provider: selectedProvider,
            apiKey: apiKey.trimmingCharacters(in: .whitespaces),
            model: finalModel,
            baseUrl: finalBaseUrl
        )
        store.systemPrompt = systemPrompt.trimmingCharacters(in: .whitespaces)
        justSaved = true
    }
}

File: dertet-harness-android/app/src/main/java/com/dertet/harness/ui/SettingsViewModel.kt
package com.dertet.harness.ui

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.dertet.harness.AppContainer
import com.dertet.harness.data.settings.ProviderType
import com.dertet.harness.network.fetchAvailableModels
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch

class SettingsViewModel(private val container: AppContainer) : ViewModel() {
    private val settingsRepo = container.settingsRepository

    var selectedProvider by mutableStateOf(ProviderType.OPENROUTER)
        private set
    var apiKey by mutableStateOf("")
    var model by mutableStateOf("")
    var baseUrl by mutableStateOf("")
    var systemPrompt by mutableStateOf("")
    var justSaved by mutableStateOf(false)
    var loaded by mutableStateOf(false)
        private set
    var availableModels by mutableStateOf<List<String>>(emptyList())
        private set
    var modelsLoading by mutableStateOf(false)
        private set

    private var modelsJob: Job? = null

    init {
        viewModelScope.launch {
            val current = settingsRepo.currentSettingsFlow().first()
            selectedProvider = current.provider
            apiKey = current.apiKey
            model = current.model.ifBlank { current.provider.defaultModel }
            baseUrl = current.baseUrl.ifBlank { current.provider.defaultBaseUrl }
            systemPrompt = current.systemPrompt
            loaded = true
            refreshModels()
        }
    }

    fun onProviderSelected(provider: ProviderType) {
        if (provider == selectedProvider) return
        selectedProvider = provider
        justSaved = false
        availableModels = emptyList()
        viewModelScope.launch {
            val s = settingsRepo.getSettingsFor(provider)
            apiKey = s.apiKey
            model = s.model.ifBlank { provider.defaultModel }
            baseUrl = s.baseUrl.ifBlank { provider.defaultBaseUrl }
            refreshModels()
        }
    }

    fun refreshModels() {
        modelsJob?.cancel()
        modelsJob = viewModelScope.launch {
            modelsLoading = true
            availableModels = fetchAvailableModels(selectedProvider.apiStyle, baseUrl, apiKey)
            modelsLoading = false
        }
    }

    fun save() {
        viewModelScope.launch {
            settingsRepo.saveProviderConfig(
                provider = selectedProvider,
                apiKey = apiKey.trim(),
                model = model.trim().ifBlank { selectedProvider.defaultModel },
                baseUrl = baseUrl.trim().ifBlank { selectedProvider.defaultBaseUrl }
            )
            settingsRepo.saveSystemPrompt(systemPrompt.trim())
            justSaved = true
        }
    }
}
